From b0726bde9b3c79675ecaf49b495d532d76517a21 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Tue, 5 Dec 2023 08:22:26 +0100
Subject: [PATCH 0146/1695] drivers: clk-stm32mp15: add the operating point
 support

MPU can operate at various clock/voltage values. A set of PLL1
configuration value can be defined. Add the functions to manage
the PLL frequency update and the associated OPP settings.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 934fd9dfeb600ef72d4cb4695b8279e3f0f82c4f drivers: clk-stm32mp15: add the operating point support]
Change-Id: I1903f4e3220dd49f65cd8aae9c3707ab4f2f981a
Signed-off-by: Etienne Carriere <etienne.carriere@st.com>
---
 core/drivers/clk/clk-stm32mp15.c         | 703 ++++++++++++++++++++++-
 core/include/drivers/stm32mp1_rcc_util.h |  11 +
 2 files changed, 713 insertions(+), 1 deletion(-)

diff --git a/core/drivers/clk/clk-stm32mp15.c b/core/drivers/clk/clk-stm32mp15.c
index 7c7cafc83..01332d702 100644
--- a/core/drivers/clk/clk-stm32mp15.c
+++ b/core/drivers/clk/clk-stm32mp15.c
@@ -31,7 +31,11 @@
 
 #include "clk-stm32-core.h"
 
+#ifdef CFG_STM32_CPU_OPP
+#define MAX_OPP		PLAT_MAX_OPP_NB
+#else
 #define MAX_OPP		(2U)
+#endif
 
 #define DT_OPP_COMPAT		"operating-points-v2"
 
@@ -648,6 +652,12 @@ static const char __maybe_unused *const stm32mp1_clk_parent_name[_PARENT_NB] = {
 	[_USB_PHY_48] = "USB_PHY_48",
 };
 
+#ifdef CFG_STM32_CPU_OPP
+/* Storage of the precomputed SoC settings for PLL1 various OPPs */
+static struct stm32mp1_pll_settings pll1_settings;
+static uint32_t current_opp_khz;
+#endif
+
 static unsigned long osc_frequency(enum stm32mp_osc_id idx)
 {
 	if (idx >= NB_OSC) {
@@ -844,6 +854,122 @@ static int pll_output(enum stm32mp1_pll_id pll_id, uint32_t output)
 	return 0;
 }
 
+#ifdef CFG_STM32_CPU_OPP
+static int pll_stop(enum stm32mp1_pll_id pll_id)
+{
+	const struct stm32mp1_clk_pll *pll = pll_ref(pll_id);
+	uint32_t pllxcr = stm32_rcc_base() + pll->pllxcr;
+	uint64_t start = 0;
+
+	/* Stop all output */
+	io_clrbits32(pllxcr, RCC_PLLNCR_DIVPEN | RCC_PLLNCR_DIVQEN |
+		     RCC_PLLNCR_DIVREN);
+
+	/* Stop PLL */
+	io_clrbits32(pllxcr, RCC_PLLNCR_PLLON);
+
+	start = timeout_init_us(PLLRDY_TIMEOUT_US);
+	/* Wait PLL stopped */
+	while (io_read32(pllxcr) & RCC_PLLNCR_PLLRDY)
+		if (timeout_elapsed(start))
+			break;
+
+	if (io_read32(pllxcr) & RCC_PLLNCR_PLLRDY) {
+		EMSG("PLL%d stop failed @ 0x%"PRIx32": 0x%"PRIx32,
+		     pll_id, pllxcr, io_read32(pllxcr));
+
+		return -1;
+	}
+
+	return 0;
+}
+
+static uint32_t pll_compute_pllxcfgr2(uint32_t *pllcfg)
+{
+	uint32_t value = 0;
+
+	value = (pllcfg[PLLCFG_P] << RCC_PLLNCFGR2_DIVP_SHIFT) &
+		RCC_PLLNCFGR2_DIVP_MASK;
+	value |= (pllcfg[PLLCFG_Q] << RCC_PLLNCFGR2_DIVQ_SHIFT) &
+		 RCC_PLLNCFGR2_DIVQ_MASK;
+	value |= (pllcfg[PLLCFG_R] << RCC_PLLNCFGR2_DIVR_SHIFT) &
+		 RCC_PLLNCFGR2_DIVR_MASK;
+
+	return value;
+}
+
+static void pll_config_output(enum stm32mp1_pll_id pll_id, uint32_t *pllcfg)
+{
+	const struct stm32mp1_clk_pll *pll = pll_ref(pll_id);
+	uintptr_t rcc_base = stm32_rcc_base();
+	uint32_t value = 0;
+
+	value = pll_compute_pllxcfgr2(pllcfg);
+
+	io_write32(rcc_base + pll->pllxcfgr2, value);
+}
+
+static int pll_compute_pllxcfgr1(const struct stm32mp1_clk_pll *pll,
+				 uint32_t *pllcfg, uint32_t *cfgr1)
+{
+	uint32_t rcc_base = stm32_rcc_base();
+	enum stm32mp1_plltype type = pll->plltype;
+	unsigned long refclk = 0;
+	uint32_t ifrge = 0;
+	uint32_t src = 0;
+
+	src = io_read32(rcc_base + pll->rckxselr) &
+	      RCC_SELR_REFCLK_SRC_MASK;
+
+	refclk = osc_frequency(pll->refclk[src]) /
+		 (pllcfg[PLLCFG_M] + 1U);
+
+	if ((refclk < (stm32mp1_pll[type].refclk_min * 1000000U)) ||
+	    (refclk > (stm32mp1_pll[type].refclk_max * 1000000U)))
+		return -1;
+
+	if ((type == PLL_800) && (refclk >= 8000000U))
+		ifrge = 1U;
+
+	*cfgr1 = (pllcfg[PLLCFG_N] << RCC_PLLNCFGR1_DIVN_SHIFT) &
+		 RCC_PLLNCFGR1_DIVN_MASK;
+	*cfgr1 |= (pllcfg[PLLCFG_M] << RCC_PLLNCFGR1_DIVM_SHIFT) &
+		  RCC_PLLNCFGR1_DIVM_MASK;
+	*cfgr1 |= (ifrge << RCC_PLLNCFGR1_IFRGE_SHIFT) &
+		  RCC_PLLNCFGR1_IFRGE_MASK;
+
+	return 0;
+}
+
+static int pll_config(enum stm32mp1_pll_id pll_id, uint32_t *pllcfg,
+		      uint32_t fracv)
+{
+	const struct stm32mp1_clk_pll *pll = pll_ref(pll_id);
+	uint32_t rcc_base = stm32_rcc_base();
+	uint32_t value = 0;
+	int ret = 0;
+
+	ret = pll_compute_pllxcfgr1(pll, pllcfg, &value);
+	if (ret)
+		return ret;
+
+	io_write32(rcc_base + pll->pllxcfgr1, value);
+
+	/* Fractional configuration */
+	io_write32(rcc_base + pll->pllxfracr, value);
+
+	/* Frac must be enabled only once its configuration is loaded */
+	value = fracv << RCC_PLLNFRACR_FRACV_SHIFT;
+	io_write32(rcc_base + pll->pllxfracr, value);
+	value = io_read32(rcc_base + pll->pllxfracr);
+	io_write32(rcc_base + pll->pllxfracr, value | RCC_PLLNFRACR_FRACLE);
+
+	pll_config_output(pll_id, pllcfg);
+
+	return 0;
+}
+#endif
+
 static unsigned long get_clock_rate(enum stm32mp1_parent_id p)
 {
 	uint32_t reg = 0;
@@ -1291,6 +1417,372 @@ void stm32mp1_clk_mcuss_protect(bool enable)
 		io_clrbits32(rcc_base + RCC_TZCR, RCC_TZCR_MCKPROT);
 }
 
+#ifdef CFG_STM32_CPU_OPP
+static void save_current_opp(void)
+{
+	unsigned long freq_khz = UDIV_ROUND_NEAREST(_stm32_clock_get_rate(CK_MPU),
+						    1000UL);
+	if (freq_khz > (unsigned long)UINT32_MAX)
+		panic();
+
+	current_opp_khz = (uint32_t)freq_khz;
+}
+
+/*
+ * Gets OPP parameters (frequency in KHz and voltage in mV) from an OPP table
+ * subnode. Platform HW support capabilities are also checked.
+ */
+static int get_opp_freqvolt_from_dt_subnode(void *fdt, int subnode,
+					    uint32_t *freq_khz,
+					    uint32_t *voltage_mv)
+{
+	const fdt64_t *cuint64 = NULL;
+	const fdt32_t *cuint32 = NULL;
+	uint64_t read_freq_64 = 0;
+	uint32_t read_voltage_32 = 0;
+
+	assert(freq_khz);
+	assert(voltage_mv);
+
+	cuint32 = fdt_getprop(fdt, subnode, "opp-supported-hw", NULL);
+	if (cuint32)
+		if (!stm32mp_supports_cpu_opp(fdt32_to_cpu(*cuint32))) {
+			DMSG("Invalid opp-supported-hw 0x%"PRIx32,
+			     fdt32_to_cpu(*cuint32));
+			return -FDT_ERR_BADVALUE;
+		}
+
+	cuint64 = fdt_getprop(fdt, subnode, "opp-hz", NULL);
+	if (!cuint64) {
+		DMSG("Missing opp-hz");
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	/* Frequency value expressed in KHz must fit on 32 bits */
+	read_freq_64 = fdt64_to_cpu(*cuint64) / 1000ULL;
+	if (read_freq_64 > (uint64_t)UINT32_MAX) {
+		DMSG("Invalid opp-hz %"PRIu64, read_freq_64);
+		return -FDT_ERR_BADVALUE;
+	}
+
+	cuint32 = fdt_getprop(fdt, subnode, "opp-microvolt", NULL);
+	if (!cuint32) {
+		DMSG("Missing opp-microvolt");
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	/* Millivolt value must fit on 16 bits */
+	read_voltage_32 = fdt32_to_cpu(*cuint32) / 1000U;
+	if (read_voltage_32 > UINT16_MAX) {
+		DMSG("Invalid opp-microvolt %"PRIu32, read_voltage_32);
+		return -FDT_ERR_BADVALUE;
+	}
+
+	*freq_khz = (uint32_t)read_freq_64;
+
+	*voltage_mv = read_voltage_32;
+
+	return 0;
+}
+
+/*
+ * Parses OPP table in DT and finds all parameters supported by the HW
+ * platform. If found, the corresponding frequency and voltage values are
+ * respectively stored in @pll1_settings structure.
+ * Note that @*count has to be set by caller to the effective size allocated
+ * for both tables. Its value is then replaced by the number of filled elements.
+ */
+static int get_all_opp_freqvolt_from_dt(uint32_t *count)
+{
+	void *fdt = NULL;
+	int node = 0;
+	int subnode = 0;
+	uint32_t idx = 0;
+
+	assert(count);
+
+	fdt = get_embedded_dt();
+	node = fdt_node_offset_by_compatible(fdt, -1, DT_OPP_COMPAT);
+	if (node < 0)
+		return node;
+
+	fdt_for_each_subnode(subnode, fdt, node) {
+		uint32_t read_freq = 0;
+		uint32_t read_voltage = 0;
+
+		if (get_opp_freqvolt_from_dt_subnode(fdt, subnode, &read_freq,
+						     &read_voltage))
+			continue;
+
+		if (idx >= *count)
+			return -FDT_ERR_NOSPACE;
+
+		pll1_settings.freq[idx] = read_freq;
+		pll1_settings.volt[idx] = read_voltage;
+		idx++;
+	}
+
+	if (!idx)
+		return -FDT_ERR_NOTFOUND;
+
+	*count = idx;
+
+	return 0;
+}
+
+static int clk_compute_pll1_settings(unsigned long input_freq, int idx)
+{
+	unsigned long post_divm = 0;
+	unsigned long long output_freq = 0;
+	unsigned long long freq = 0;
+	unsigned long long vco = 0;
+	int divm = 0;
+	int divn = 0;
+	int divp = 0;
+	int frac = 0;
+	int i = 0;
+	unsigned int diff = 0;
+	unsigned int best_diff = UINT_MAX;
+
+	output_freq = (unsigned long long)pll1_settings.freq[idx] * 1000U;
+
+	/* Following parameters have always the same value */
+	pll1_settings.cfg[idx][PLLCFG_Q] = 0;
+	pll1_settings.cfg[idx][PLLCFG_R] = 0;
+	pll1_settings.cfg[idx][PLLCFG_O] = PQR(1, 0, 0);
+
+	for (divm = DIVM_MAX; divm >= DIVM_MIN; divm--)	{
+		post_divm = input_freq / (unsigned long)(divm + 1);
+
+		if ((post_divm < POST_DIVM_MIN) ||
+		    (post_divm > POST_DIVM_MAX))
+			continue;
+
+		for (divp = DIVP_MIN; divp <= DIVP_MAX; divp++) {
+
+			freq = output_freq * (divm + 1) * (divp + 1);
+
+			divn = (int)((freq / input_freq) - 1);
+			if ((divn < DIVN_MIN) || (divn > DIVN_MAX))
+				continue;
+
+			frac = (int)(((freq * FRAC_MAX) / input_freq) -
+				     ((divn + 1) * FRAC_MAX));
+
+			/* 2 loops to refine the fractional part */
+			for (i = 2; i != 0; i--) {
+				if (frac > FRAC_MAX)
+					break;
+
+				vco = (post_divm * (divn + 1)) +
+				      ((post_divm * (unsigned long long)frac) /
+				       FRAC_MAX);
+
+				if ((vco < (VCO_MIN / 2)) ||
+				    (vco > (VCO_MAX / 2))) {
+					frac++;
+					continue;
+				}
+
+				freq = vco / (divp + 1);
+				if (output_freq < freq)
+					diff = (unsigned int)(freq -
+							      output_freq);
+				else
+					diff = (unsigned int)(output_freq -
+							      freq);
+
+				if (diff < best_diff)  {
+					pll1_settings.cfg[idx][PLLCFG_M] = divm;
+					pll1_settings.cfg[idx][PLLCFG_N] = divn;
+					pll1_settings.cfg[idx][PLLCFG_P] = divp;
+					pll1_settings.frac[idx] = frac;
+
+					if (!diff)
+						return 0;
+
+					best_diff = diff;
+				}
+
+				frac++;
+			}
+		}
+	}
+
+	if (best_diff == UINT_MAX) {
+		pll1_settings.cfg[idx][PLLCFG_O] = 0;
+		return -1;
+	}
+
+	return 0;
+}
+
+static int clk_get_pll1_settings(uint32_t clksrc, int index)
+{
+	unsigned long input_freq = 0;
+	unsigned int i = 0;
+
+	for (i = 0; i < MAX_OPP; i++)
+		if (pll1_settings.freq[i] == pll1_settings.freq[index])
+			break;
+
+	if (((i == MAX_OPP) &&
+	     !stm32mp1_clk_pll1_settings_are_valid()) ||
+	    ((i < MAX_OPP) && !pll1_settings.cfg[i][PLLCFG_O])) {
+		/*
+		 * Either PLL1 settings structure is completely empty,
+		 * or these settings are not yet computed: do it.
+		 */
+		switch (clksrc) {
+		case CLK_PLL12_HSI:
+			input_freq = _stm32_clock_get_rate(CK_HSI);
+			break;
+		case CLK_PLL12_HSE:
+			input_freq = _stm32_clock_get_rate(CK_HSE);
+			break;
+		default:
+			panic();
+		}
+
+		return clk_compute_pll1_settings(input_freq, index);
+	}
+
+	if (i < MAX_OPP) {
+		if (pll1_settings.cfg[i][PLLCFG_O])
+			return 0;
+
+		/*
+		 * Index is in range and PLL1 settings are computed:
+		 * use content to answer to the request.
+		 */
+		memcpy(&pll1_settings.cfg[index][0], &pll1_settings.cfg[i][0],
+		       sizeof(uint32_t) * PLLCFG_NB);
+		pll1_settings.frac[index] = pll1_settings.frac[i];
+
+		return 0;
+	}
+
+	return -1;
+}
+
+static int clk_save_current_pll1_settings(uint32_t buck1_voltage)
+{
+	const struct stm32mp1_clk_pll *pll = pll_ref(_PLL1);
+	uint32_t rcc_base = stm32_rcc_base();
+	uint32_t freq = 0;
+	unsigned int i = 0;
+
+	freq = UDIV_ROUND_NEAREST(_stm32_clock_get_rate(CK_MPU), 1000L);
+
+	for (i = 0; i < MAX_OPP; i++)
+		if (pll1_settings.freq[i] == freq)
+			break;
+
+	if ((i == MAX_OPP) ||
+	    ((pll1_settings.volt[i] != buck1_voltage) && buck1_voltage))
+		return -1;
+
+	pll1_settings.cfg[i][PLLCFG_M] = (io_read32(rcc_base + pll->pllxcfgr1) &
+					  RCC_PLLNCFGR1_DIVM_MASK) >>
+					 RCC_PLLNCFGR1_DIVM_SHIFT;
+
+	pll1_settings.cfg[i][PLLCFG_N] = (io_read32(rcc_base + pll->pllxcfgr1) &
+					  RCC_PLLNCFGR1_DIVN_MASK) >>
+					 RCC_PLLNCFGR1_DIVN_SHIFT;
+
+	pll1_settings.cfg[i][PLLCFG_P] = (io_read32(rcc_base + pll->pllxcfgr2) &
+					  RCC_PLLNCFGR2_DIVP_MASK) >>
+					 RCC_PLLNCFGR2_DIVP_SHIFT;
+
+	pll1_settings.cfg[i][PLLCFG_Q] = (io_read32(rcc_base + pll->pllxcfgr2) &
+					  RCC_PLLNCFGR2_DIVQ_MASK) >>
+					 RCC_PLLNCFGR2_DIVQ_SHIFT;
+
+	pll1_settings.cfg[i][PLLCFG_R] = (io_read32(rcc_base + pll->pllxcfgr2) &
+					  RCC_PLLNCFGR2_DIVR_MASK) >>
+					 RCC_PLLNCFGR2_DIVR_SHIFT;
+
+	pll1_settings.cfg[i][PLLCFG_O] = io_read32(rcc_base + pll->pllxcr) >>
+					 RCC_PLLNCR_DIVEN_SHIFT;
+
+	pll1_settings.frac[i] = (io_read32(rcc_base + pll->pllxfracr) &
+				 RCC_PLLNFRACR_FRACV_MASK) >>
+				RCC_PLLNFRACR_FRACV_SHIFT;
+
+	return i;
+}
+
+static uint32_t stm32mp1_clk_get_pll1_current_clksrc(void)
+{
+	uint32_t value = 0;
+	const struct stm32mp1_clk_pll *pll = pll_ref(_PLL1);
+	uint32_t rcc_base = stm32_rcc_base();
+
+	value = io_read32(rcc_base + pll->rckxselr);
+
+	switch (value & RCC_SELR_REFCLK_SRC_MASK) {
+	case 0:
+		return CLK_PLL12_HSI;
+	case 1:
+		return CLK_PLL12_HSE;
+	default:
+		panic();
+	}
+}
+
+int stm32mp1_clk_compute_all_pll1_settings(uint32_t buck1_voltage)
+{
+	unsigned int i = 0;
+	int ret = 0;
+	int index = 0;
+	uint32_t count = MAX_OPP;
+	uint32_t clksrc = 0;
+
+	ret = get_all_opp_freqvolt_from_dt(&count);
+	switch (ret) {
+	case 0:
+		break;
+	case -FDT_ERR_NOTFOUND:
+		DMSG("Cannot find all OPP info in DT: use default settings.");
+		return 0;
+	default:
+		EMSG("Inconsistent OPP settings found in DT, ignored.");
+		return 0;
+	}
+
+	index = clk_save_current_pll1_settings(buck1_voltage);
+
+	clksrc = stm32mp1_clk_get_pll1_current_clksrc();
+
+	for (i = 0; i < count; i++) {
+		if (index >= 0 && i == (unsigned int)index)
+			continue;
+
+		ret = clk_get_pll1_settings(clksrc, i);
+		if (ret != 0)
+			return ret;
+	}
+
+	pll1_settings.valid_id = PLL1_SETTINGS_VALID_ID;
+
+	return 0;
+}
+
+void stm32mp1_clk_lp_save_opp_pll1_settings(uint8_t *data, size_t size)
+{
+	if ((size != sizeof(pll1_settings)) ||
+	    !stm32mp1_clk_pll1_settings_are_valid())
+		panic();
+
+	memcpy(data, &pll1_settings, size);
+}
+
+bool stm32mp1_clk_pll1_settings_are_valid(void)
+{
+	return pll1_settings.valid_id == PLL1_SETTINGS_VALID_ID;
+}
+#endif
+
 /*
  * Conversion between clk references and clock gates and clock on internals
  *
@@ -2985,7 +3477,9 @@ static TEE_Result stm32mp1_clock_provider_probe(const void *fdt, int offs,
 	}
 
 	enable_static_secure_clocks();
-
+#ifdef CFG_STM32_CPU_OPP
+	save_current_opp();
+#endif
 	register_pm_core_service_cb(stm32_clock_pm, NULL,
 				    "stm32mp15-clk-service");
 
@@ -3005,6 +3499,210 @@ DEFINE_DT_DRIVER(stm32mp1_clock_dt_driver) = {
 	.probe = stm32mp1_clock_provider_probe,
 };
 
+#ifdef CFG_STM32_CPU_OPP
+/* Start MPU OPP */
+
+/*
+ * Check if PLL1 can be configured on the fly.
+ * @result  (-1) => config on the fly is not possible.
+ *          (0)  => config on the fly is possible.
+ *          (+1) => same parameters as those in place, no need to reconfig.
+ * Return value is 0 if no error.
+ */
+static int is_pll_config_on_the_fly(enum stm32mp1_pll_id pll_id,
+				    uint32_t *pllcfg, uint32_t fracv,
+				    int *result)
+{
+	const struct stm32mp1_clk_pll *pll = pll_ref(pll_id);
+	uintptr_t rcc_base = stm32_rcc_base();
+	uint32_t fracr = 0;
+	uint32_t value = 0;
+	int ret = 0;
+
+	ret = pll_compute_pllxcfgr1(pll, pllcfg, &value);
+	if (ret)
+		return ret;
+
+	if (io_read32(rcc_base + pll->pllxcfgr1) != value) {
+		/* Different DIVN/DIVM, can't config on the fly */
+		*result = -1;
+		return 0;
+	}
+
+	*result = true;
+
+	fracr = fracv << RCC_PLLNFRACR_FRACV_SHIFT;
+	fracr |= RCC_PLLNFRACR_FRACLE;
+	value = pll_compute_pllxcfgr2(pllcfg);
+
+	if ((io_read32(rcc_base + pll->pllxfracr) == fracr) &&
+	    (io_read32(rcc_base + pll->pllxcfgr2) == value))
+		/* Same parameters, no need to config */
+		*result = 1;
+	else
+		*result = 0;
+
+	return 0;
+}
+
+static int stm32mp1_get_mpu_div(uint32_t freq_khz)
+{
+	unsigned long freq_pll1_p;
+	unsigned long div;
+
+	freq_pll1_p = get_clock_rate(_PLL1_P) / 1000UL;
+	if ((freq_pll1_p % freq_khz) != 0U)
+		return -1;
+
+	div = freq_pll1_p / freq_khz;
+
+	switch (div) {
+	case 1UL:
+	case 2UL:
+	case 4UL:
+	case 8UL:
+	case 16UL:
+		return __builtin_ffs(div) - 1;
+	default:
+		return -1;
+	}
+}
+
+/* Configure PLL1 from input frequency OPP parameters */
+static int pll1_config_from_opp_khz(uint32_t freq_khz)
+{
+	unsigned int idx = 0;
+	int ret = 0;
+	int div = 0;
+	int config_on_the_fly = -1;
+
+	for (idx = 0; idx < MAX_OPP; idx++)
+		if (pll1_settings.freq[idx] == freq_khz)
+			break;
+
+	if (idx == MAX_OPP)
+		return -1;
+
+	div = stm32mp1_get_mpu_div(freq_khz);
+	switch (div) {
+	case -1:
+		break;
+	case 0:
+		return stm32_clk_configure_mux(CLK_MPU_PLL1P);
+	default:
+		if (!stm32_div_set_value(DIV_MPU, div))
+			return stm32_clk_configure_mux(CLK_MPU_PLL1P_DIV);
+
+		return -1;
+	}
+
+	ret = is_pll_config_on_the_fly(_PLL1, &pll1_settings.cfg[idx][0],
+				       pll1_settings.frac[idx],
+				       &config_on_the_fly);
+	if (ret)
+		return ret;
+
+	if (config_on_the_fly == 1)
+		return 0;
+
+	if (config_on_the_fly == -1) {
+		/* Switch to HSI and stop PLL1 before reconfiguration */
+		ret = stm32_clk_configure_mux(CLK_MPU_HSI);
+		if (ret)
+			return ret;
+
+		ret = pll_stop(_PLL1);
+		if (ret)
+			return ret;
+	}
+
+	ret = pll_config(_PLL1, &pll1_settings.cfg[idx][0],
+			 pll1_settings.frac[idx]);
+	if (ret)
+		return ret;
+
+	if (config_on_the_fly == -1) {
+		/* Start PLL1 and switch back to after reconfiguration */
+		pll_start(_PLL1);
+
+		ret = pll_output(_PLL1, pll1_settings.cfg[idx][PLLCFG_O]);
+		if (ret)
+			return ret;
+
+		ret = stm32_clk_configure_mux(CLK_MPU_PLL1P);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+TEE_Result stm32mp1_set_opp_khz(uint32_t freq_khz)
+{
+	uint32_t mpu_src = 0;
+
+	if (freq_khz == current_opp_khz)
+		return TEE_SUCCESS;
+
+	if (!stm32mp1_clk_pll1_settings_are_valid()) {
+		/*
+		 * No OPP table in DT or an error occurred during PLL1
+		 * settings computation, system can only work on current
+		 * operating point so return error.
+		 */
+		return TEE_ERROR_NO_DATA;
+	}
+
+	/* Check that PLL1 is MPU clock source */
+	mpu_src = io_read32(stm32_rcc_base() + RCC_MPCKSELR) &
+		RCC_SELR_SRC_MASK;
+	if ((mpu_src != RCC_MPCKSELR_PLL) &&
+	    (mpu_src != RCC_MPCKSELR_PLL_MPUDIV))
+		return TEE_ERROR_BAD_STATE;
+
+	if (pll1_config_from_opp_khz(freq_khz)) {
+		/* Restore original value */
+		if (pll1_config_from_opp_khz(current_opp_khz)) {
+			EMSG("No CPU operating point can be set");
+			panic();
+		}
+
+		return TEE_ERROR_GENERIC;
+	}
+
+	current_opp_khz = freq_khz;
+
+	return TEE_SUCCESS;
+}
+
+int stm32mp1_round_opp_khz(uint32_t *freq_khz)
+{
+	unsigned int i = 0;
+	uint32_t round_opp = 0;
+
+	if (!stm32mp1_clk_pll1_settings_are_valid()) {
+		/*
+		 * No OPP table in DT, or an error occurred during PLL1
+		 * settings computation, system can only work on current
+		 * operating point, so return current CPU frequency.
+		 */
+		*freq_khz = current_opp_khz;
+
+		return 0;
+	}
+
+	for (i = 0; i < MAX_OPP; i++)
+		if ((pll1_settings.freq[i] <= *freq_khz) &&
+		    (pll1_settings.freq[i] > round_opp))
+			round_opp = pll1_settings.freq[i];
+
+	*freq_khz = round_opp;
+
+	return 0;
+}
+/* End MPU OPP */
+#endif
+
 struct soc_stop_context {
 	uint32_t pll3cr;
 	uint32_t pll4cr;
@@ -3363,6 +4061,9 @@ static TEE_Result stm32_clock_pm(enum pm_op op, unsigned int pm_hint __unused,
 	if (op == PM_OP_SUSPEND) {
 		stm32_clock_suspend();
 	} else {
+#ifdef CFG_STM32_CPU_OPP
+		pll1_config_from_opp_khz(current_opp_khz);
+#endif
 		stm32_clock_resume();
 	}
 
diff --git a/core/include/drivers/stm32mp1_rcc_util.h b/core/include/drivers/stm32mp1_rcc_util.h
index 7456d8ccc..c53d38908 100644
--- a/core/include/drivers/stm32mp1_rcc_util.h
+++ b/core/include/drivers/stm32mp1_rcc_util.h
@@ -37,6 +37,17 @@ void stm32mp_register_clock_parents_secure(unsigned long clock_id __unused)
 /* Protect the MCU clock subsytem */
 void stm32mp1_clk_mcuss_protect(bool enable);
 
+#ifdef CFG_STM32_CPU_OPP
+/*
+ * Util for PLL1 settings management based on DT OPP table content.
+ */
+int stm32mp1_clk_compute_all_pll1_settings(uint32_t buck1_voltage);
+void stm32mp1_clk_lp_save_opp_pll1_settings(uint8_t *data, size_t size);
+bool stm32mp1_clk_pll1_settings_are_valid(void);
+TEE_Result stm32mp1_set_opp_khz(uint32_t freq_khz);
+int stm32mp1_round_opp_khz(uint32_t *freq_khz);
+#endif
+
 /* PM sequences specific to SoC STOP mode support */
 void stm32mp1_clk_save_context_for_stop(void);
 void stm32mp1_clk_restore_context_for_stop(void);
-- 
2.47.2

